### PR TITLE
[clang] Implement __builtin_popcountg

### LINKED ISSUE #82058: [clang] implement __builtin_popcountg
GCC recently added this builtin <https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fpopcountg> which makes implementing C23's `stdc_count_ones` (and `stdc_count_zeros`) way way simpler.  In particular, because it's type generic, it makes it easier to support not only the existing standard types pre-C23, but the extended 128b+ types (`__int128`) and C23 addition `BitInt` types.

This was alluded to by @pinskia in https://github.com/llvm/llvm-project/issues/79630#issuecomment-1912561197.

### PR DESCRIPTION
**Pull request by overmighty**
Fixes #82058.

### CODE REVIEW

**Comment by overmighty** [comment]
Should I add support for `__builtin_popcountg` in constant expressions in this PR too or in a later one?

**Comment by efriedma-quic** [comment]
Missing semantic analysis.  Since the signature is unspecified in Builtin.td, you have to check that there's one argument, and that argument is an integer.  That code should go in SemaChecking.cpp.

You can leave constant evaluation to a followup, sure.

**Review by nickdesaulniers — COMMENTED** [review]
Awesome, thanks!

Don't forget to update https://clang.llvm.org/docs/LanguageExtensions.html#builtin-functions

**Review by RKSimon — COMMENTED** [review]
Please can you add constexpr test coverage?

**Comment by overmighty** [comment]
@RKSimon The builtin currently can't be used with `constexpr`. Support for constant evaluation is planned for a follow-up PR unless you would like me to add it in this one. Should I remove the `Constexpr` attribute from the builtin in Builtins.td for now?

**Comment by RKSimon** [comment]
> @RKSimon The builtin currently can't be used with `constexpr`. Support for constant evaluation is planned for a follow-up PR unless you would like me to add it in this one. Should I remove the `Constexpr` attribute from the builtin in Builtins.td for now?

Yes please!

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by overmighty** [comment]
Done.

**Comment by nickdesaulniers** [comment]
Perhaps worth rebasing then force pushing to this branch to rekick the build bots?

Thanks for implementing this @overmighty !  There's a few more builtins like this that would simplify implementations of more C23 library additions. Let me know if you're interested in those, too.

**Comment by overmighty** [comment]
Rebased.

@nickdesaulniers I am interested in implementing more builtins like this one. :)

**Comment by nickdesaulniers** [comment]
Looks like the linter is upset with pre-existing changes.  ok2land. Are you able to merge this @overmighty or do you need us to?

**Comment by overmighty** [comment]
@nickdesaulniers I need someone to merge this for me, I don't have write/push access. Thanks.

**Comment by nickdesaulniers** [comment]
> @nickdesaulniers I am interested in implementing more builtins like this one. :)


Cool, I've filed:
1. https://github.com/llvm/llvm-project/issues/83075
2. https://github.com/llvm/llvm-project/issues/83076

Those would be helpful to have next, and I suspect as straight foward to implement as popcountg was.

**Comment by overmighty** [comment]
Maybe the semantic analysis and docs should have been changed to match GCC's better.

GCC's docs say their `__builtin_popcountg` takes a "type-generic unsigned integer" but that "No integral argument promotions are performed on the argument." Our builtin accepts signed integers too, but the argument goes through integer promotion, so the builtin can return an incorrect value if we give it a signed integer.

For example, this program will output 32 with Clang on `x86_64-unknown-linux-gnu`:

```c
#include <stdio.h>

int main(void) {
    char foo = -1;
    int pop = __builtin_popcountg(foo);
    printf("%d\n", pop);
}
```

GCC refuses to compile it: https://godbolt.org/z/chh4WGT4v.

Should I open a new issue?

**Comment by overmighty** [comment]
Thanks. I plan to fix it soon, so I will probably just open a PR then.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -3216,7 +3216,8 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
   case Builtin::BI__popcnt64:
   case Builtin::BI__builtin_popcount:
   case Builtin::BI__builtin_popcountl:
-  case Builtin::BI__builtin_popcountll: {
+  case Builtin::BI__builtin_popcountll:
+  case Builtin::BI__builtin_popcountg: {
```
Comment (bjope) [inline clang/lib/CodeGen/CGBuiltin.cpp:3220]: I've opened a pull-request to also adjust the codegen here, to make sure that we use an unsigned cast between the LLVM intrinsic result type and the builtins return type. See https://github.com/llvm/llvm-project/pull/90845

**Pull request merged**